enumerable.rb

```diff
@@ -0,0 +1,141 @@
+# reimplementing methods
+private
+
+def validate?(arg, element, equality)
+  if arg.instance_of? Regexp
+    arg.match?(element.to_s)
+  elsif element.instance_of? arg.class
+    arg == element
+  elsif arg.instance_of? Class
+    element.is_a? arg
+  else
+    equality
+  end
+end
+
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    size.times { |i| yield(to_a[i]) }
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    size.times { |i| yield(to_a[i], i) }
+    self
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    arr = []
+    my_each { |i| arr << i if yield(i) }
+    arr
+  end
+
+  def my_all?(arg = nil)
+    my_each do |i|
+      condition = block_given? && !yield(i)
+      condition ||= !block_given? && arg.nil? && !i
+      condition ||= !arg.nil? && !validate?(arg, i, false)
+      return false if condition
+    end
+    true
+  end
+
+  def my_any?(arg = nil)
+    my_each do |i|
+      condition = block_given? && yield(i)
+      condition ||= !block_given? && arg.nil? && i
+      condition ||= !arg.nil? && validate?(arg, i, false)
+      return true if condition
+    end
+    false
+  end
+
+  def my_none?(arg = nil)
+    my_each do |i|
+      condition = block_given? && yield(i)
+      condition ||= !block_given? && arg.nil? && i
+      condition ||= !arg.nil? && validate?(arg, i, false)
+      return false if condition
+    end
+    true
+  end
+
+  def my_count(arg = nil)
+    sum = 0
+    my_each do |i|
+      condition = arg.nil? && !block_given?
+      condition ||= block_given? && yield(i)
+      condition ||= !arg.nil? && i == arg
+      sum += 1 if condition
+    end
+    sum
+  end
+
+  def my_map
+    return to_enum(:my_map) unless block_given?
+
+    arr = []
+    my_each { |i| arr << yield(i) }
+    arr
+  end
+
+  def inject_arg_valid?(*arg)
+    acc = arg[0]&.is_a?(Numeric) ? arg[0] : to_a[0]
+    if arg[0]&.is_a? Symbol
+      operation = arg[0]
+    elsif arg[1]&.is_a? Symbol
+      operation = arg[1]
+    end
+    [acc, operation]
+  end
+
+  # awesomeness in each line
+  def eval_operation?(operation)
+    ':/ :% :-'.include?(operation.to_s)
+  end
+
+  def inject_valid_symbol?
+    lambda do |arg, i|
+      res = (arg.size == 1 && !i.zero?)
+      res ||= arg.size == 2
+      res
+    end
+  end
+
+  def next_index?(arg, index)
+    ret = arg.size == 2 || arg.size.zero?
+    ret &&= index.zero?
+    ret
+  end
+
+  def check_operation(operation)
+    [acc, e] if eval_operation?(operation)
```

Comment:
What's the return value in case the condition is false? Where are acc and e coming from?

```ruby
  def check_operation(operation, acc, nxt)
    eval_operation?(operation) ? [acc, nxt] : [nxt, acc]
  end
```